Code editor improvements: refactoring, selection handling, and new commands

maxland255 opened the pull request:
This pull request introduces several enhancements to the code editor, aimed at modernizing the existing code and improving the user experience.

- **Variable declaration refactoring:**
  - Replaced `var` with `const` or `let` where necessary to follow modern JavaScript best practices. This also adds an extra layer of security by ensuring that variables are block-scoped and not unintentionally redefined.

- **Conditional syntax improvement:**
  - Updated the `switch` statements to a more modern `match` syntax:
    - New syntax: `match (VARIABLE) \n case (VALUE) \n default \n endmatch`.
    - The alternative `if`, `else if`, `else`, `endif` syntax remains functional but has been adapted to a more conventional structure.

- **New editing features:**
  - Text selection now allows for automatic addition of quotes, parentheses, etc., around the selected text.
  - Multi-line selection with the ability to adjust indentation using `Tab` and `Shift+Tab`.

- **New text scaling commands:**
  - Added shortcuts `Ctrl+` and `Ctrl-` to increase or decrease text size.
  - Shortcut `Ctrl0` to reset text size to its default value.

- **Updated JavaScript functions:**
  - Fixed obsolete functions and used newer methods where applicable.

These changes bring improvements in performance, readability, and user convenience.

gaetanledocte commented on scripts/app.js:
```diff
@@ -115,16 +166,79 @@ input.addEventListener("keyup", function() {
+    }
+
+    // Parenthesis
+    if (event.key === '(' || event.key === '[' || event.key === '{' || event.key === '"' || event.key === "'") {
```
You can map `open` and `close` characters based on a dictionary, to improve the maintainability of the code (we can more easily add elements):

```
const closingBlocks = { // the name can be challenged
    "(": ")",
    "[": "]",
    "{": "}",
    "'": "'",
}

if (event.key in closingBlocks) {
    // ...

    const close = closingBlocks[event.key];
}
```

gaetanledocte commented on scripts/app.js:
```diff
@@ -89,8 +93,55 @@ document.addEventListener("keydown", function(event) {
+    if (event.ctrlKey && event.code === "Slash") {
+        event.preventDefault();
+        zoom(5);
+    }else if (event.ctrlKey && event.code === "Equal") {
```
Add a space between `}` and `else` ?

gaetanledocte commented on scripts/app.js:
```diff
@@ -89,8 +93,55 @@ document.addEventListener("keydown", function(event) {
+    }else if (event.ctrlKey && event.code === "Equal") {
+        event.preventDefault();
+        zoom(-5);
+    }else if (event.ctrlKey && event.code === "Digit0") {
```
Add a space between `}` and `else` ?

gaetanledocte commented on scripts/PseudoCodeParser.js:
```diff
@@ -1,42 +1,46 @@
+        { pattern: /\benddo\b/i,        replacement: "╙──",         border: false },
+        { pattern: /\bendwhile\b/i,     replacement: "╙──",         border: false },
+        { pattern: /===+/i,             replacement: "╙──",         border: false },
+        { pattern: /\bmatch\b/i,        replacement: "┌── match",  border: "│" },
```
Add an extra spacing before `border`, to fix alignment.

gaetanledocte commented on scripts/app.js:
```diff
@@ -89,8 +93,55 @@ document.addEventListener("keydown", function(event) {
 
         saveContentToCache();
     }
+
```
We can isolate the condition on `event.ctrlKey`, to avoid repeating it:

```
if (event.ctrlKey) {
    if (event.code === "Slash") {
        ...
    } else if (...) {
        ...
    }
```

And, given the current structure of the code, a guard clauses pattern can even be used:

```
if (!event.ctrlKey) {
    return;
}

if (event.code === "Slash") {
    ...
} else if (...) {
    ...
}
```

gaetanledocte commented on scripts/app.js:
```diff
@@ -89,8 +93,55 @@ document.addEventListener("keydown", function(event) {
+
+function zoom(zoomValue) {
+    const currentFontSize = parseInt(window.getComputedStyle(input).fontSize);
+
```
`currentFontSize + zoomValue` is repeated 4 times in the function: you can make it a constant.

```
const newFontSize = currentFontSize + zoomValue
```

gaetanledocte commented on scripts/app.js:
```diff
@@ -89,8 +93,55 @@ document.addEventListener("keydown", function(event) {
+    resetZoom();
+});
+
+function zoom(zoomValue) {
```
To challenge a little:

- in the project, the font size is historically expressed in relative units (rem) and not in absolute units (px): it could be good to keep the notion of relative size
- on my screen, the base font size is 12.8px, which means that, if I understand the code correctly, zooming out is impossible and the maximum possible zoom is 37.8px (all possible values ​​in my case being: 12.8px, 17.8px, 22.8px, 27.8px, 32.8px and 37.8px)
The zoom feature as it stands remains interesting, but could be further improved in the future, if needed.

maxland255 replied:
Thank you for your feedback!

I’ve implemented a solution that works with rem instead of px to ensure the font size is relative. Here’s the approach I took:

```js
const fontSizeRem = parseFloat(window.getComputedStyle(input).fontSize) / parseFloat(window.getComputedStyle(document.documentElement).fontSize);
const lineHeightRem = parseFloat(window.getComputedStyle(input).lineHeight) / parseFloat(window.getComputedStyle(document.documentElement).fontSize);

const newFontSizeRem = fontSizeRem + zoomValue;
const newLineHeightRem = lineHeightRem + zoomValue;

if (newFontSizeRem < 0.5 || newFontSizeRem > 2) {
    return;
}

input.style.fontSize = `${newFontSizeRem}rem`;
output.style.fontSize = `${newFontSizeRem}rem`;
input.style.lineHeight = `${newLineHeightRem}rem`;
output.style.lineHeight = `${newLineHeightRem}rem`;
```

This approach should now allow zooming in and out while maintaining relative font sizing, as you suggested. The zoom range is currently set between 0.5rem and 2rem, but it can be adjusted if needed. Let me know what you think or if further adjustments are required!

gaetanledocte replied:
Looks good to me! 😄 Thank you!

gaetanledocte commented on scripts/app.js:
```diff
@@ -115,16 +166,79 @@ input.addEventListener("keyup", function() {
+    if (event.code === "Tab") {
+        event.preventDefault();
+
+        manageTabulation(2);
```
Add `return;` ?

gaetanledocte commented on scripts/app.js:
```diff
@@ -115,16 +166,79 @@ input.addEventListener("keyup", function() {
+        const open = event.key;
+        const close = (event.key === '(') ? ')' : (event.key === '[') ? ']' : (event.key === '{') ? '}' : event.key;
+
+        // Set textarea value to: text before selection + open + selected text + close + text after selection
```
Comment is useless and redondant with the code.

If you want, you can add named variables for `this.value.substring(0, start)` and `this.value.substring(end)`, and you'll obtain the same readibility as with your comment.

gaetanledocte commented on scripts/app.js:
```diff
@@ -281,10 +395,18 @@ function saveContentToCache(){
+    input.value = localStorage['codeDA'] || '';
 }
 
 function isSaveKeyboardShortcut(event) {
```
Here is an idea of implementation to prevent redondant `if` statements:

```
const userAgent = window.navigator.userAgent;

const osMapping = {
    "Linux": "Linux",
    "Mac": "Mac",
    "Win": "Windows",
    "X11": "Android"
};

const os = Object.keys(osMapping).find(key => userAgent.includes(key)) || "unknown";

return (os === "Mac" ? event.metaKey : event.ctrlKey) && event.keyCode === 83;
```

gaetanledocte reviewed (commented):
Overall, LGTM. 👏

HunteRoi reviewed (changes requested):
First of all, thank you very much for the time you took in updating the tool! 
It has been a long time since such an addition was last made and it is very heartwarming to see your interest in this.

However, as I always have interoperability in mind, I hope you understand that I cannot merge your PR as is. Fortunately, the only changes I would like you to make is _not to remove the if-elseif-else_ pattern because many previously made DAs are relying on such. I would advise to first make an attempt for both match and if-else if-else patterns to coexist in a non-breaking manner before a next release where the latest would disappear.

I understand this may be a pain to update but I think this would be a less abrupt move for the end users.

Again, thank you for your work!

maxland255 commented:
I have implemented the requested changes from the review. Specifically, I have:

- Updated the zoom function to use relative units (`rem`) instead of absolute units (`px`) for better scalability.
- Reintroduced compatibility with the old switch syntax (`if`, `elseif`, `else`, `endif`) while maintaining the new syntax, allowing both to coexist.
- Made minor code adjustments as discussed, including adding spaces before else statements and other formatting improvements.

Please take a look at the latest commits and let me know if there are any further adjustments needed. Thank you for your feedback!

gaetanledocte commented on scripts/app.js:
```diff
@@ -123,16 +127,22 @@ buttonZoomReset.addEventListener("click", function(event) {
+    const newLineHeightRem = lineHeightRem + zoomValue;
+    const newOutputLineHeightRem = outputLineHeightRem + zoomValue;
+
+    if (newFontSizeRem < 0.5 || newFontSizeRem > 2) {
```
This is really minor, but you can move this condition before the initialization of all constants: currently, you are initializing quite a few constants, which might never be used.

```
const fontSizeRem = parseFloat(window.getComputedStyle(input).fontSize) / documentFontSize;
const newFontSizeRem = fontSizeRem + zoomValue;

if (newFontSizeRem < 0.5 || newFontSizeRem > 2) {
    return;
}

const lineHeightRem = parseFloat(window.getComputedStyle(input).lineHeight) / documentFontSize;
const outputLineHeightRem = parseFloat(window.getComputedStyle(output).lineHeight) / documentFontSize;

const newLineHeightRem = lineHeightRem + zoomValue;
const newOutputLineHeightRem = outputLineHeightRem + zoomValue;

...
```

gaetanledocte commented on scripts/app.js:
```diff
@@ -123,16 +127,22 @@ buttonZoomReset.addEventListener("click", function(event) {
-    const currentFontSize = parseInt(window.getComputedStyle(input).fontSize);
+    const fontSizeRem = parseFloat(window.getComputedStyle(input).fontSize) / parseFloat(window.getComputedStyle(document.documentElement).fontSize);
+    const lineHeightRem = parseFloat(window.getComputedStyle(input).lineHeight) / parseFloat(window.getComputedStyle(document.documentElement).fontSize);
+    const outputLineHeightRem = parseFloat(window.getComputedStyle(output).lineHeight) / parseFloat(window.getComputedStyle(document.documentElement).fontSize);
```
`parseFloat(window.getComputedStyle(document.documentElement).fontSize)` est répété souvent : tu peux l'extraire dans une constante.

`const documentFontSize = parseFloat(window.getComputedStyle(document.documentElement).fontSize);`

maxland255 replied:
I made the changes you suggested in the last commit.

gaetanledocte reviewed (commented):
Thank you @maxland255!

Here is some minor comment on your iteration. 

Otherwize, PR can be merged for me. 😉

maxland255 commented:
I have reviewed the change requests once more to ensure that I didn’t miss anything, and I believe all requested modifications have now been made. The changes are included in the last commit. If there’s anything else that needs attention, please feel free to let me know.
Thank you again for your time and feedback!

HunteRoi commented:
There is a minor styling issue due to the number of buttons in the taskbar for certain screen sizes but otherwise, the PR can be merged for me.
![image](https://github.com/user-attachments/assets/610e3707-560a-4b74-b49c-b88d877eccb2)

maxland255 commented:
I agree with you, that’s why I made it scrollable. But if you have a better idea or a different placement for the zoom buttons, I can change that. Maybe I could place them near the result instead of where the code is written?

HunteRoi commented:
> I agree with you, that’s why I made it scrollable. 

As far as I am concerned, the horizontal scrollbar does not get enabled.

> But if you have a better idea or a different placement for the zoom buttons, I can change that. Maybe I could place them near the result instead of where the code is written?

Semantically speaking, you could move the zoom-related buttons to the header navbar or keep them in the main container. But I think that the whole menu should be reworked and this should take place in another PR.

@gaetanledocte any thought on this last bit? ^

maxland255 commented:
Alright, I won’t change anything for now, so I’ll leave the zoom buttons as they are.
Thank you for your feedback!
